File: height balanced.cpp
// height balanced subtree
#include<bits/stdc++.h>
using namespace std;


struct node
{
	int data;
	struct node*left;
	struct node*right;
	node(int data){
		this->data=data;
		this->left=NULL;
		this->right=NULL;
	}
	
};


pair<int,bool> heightBalanced(struct node*root){
	
	
	pair<int,bool> p;
	
	if(root==NULL)
	{
		p.first=0;
		p.second=true;
		return p;
	}
	
	
	pair<int,bool> left=heightBalanced(root->left);
	pair<int,bool> right=heightBalanced(root->right);
	
	
	p.first=max(left.first,right.first)+1;
	
	if(abs(left.first - right.first)<=1 && left.second==true && right.second==true){
	p.second=true;	
	}else
	p.second=false;
	
	return p;
	
	//lsbt is heeight balanced
	
	// rsbt is height balanced
	
	
	// difference in height of lssbt and rsbt is 1
	
	
	
	
	
}

int main(){
	
	
	
}

File: src/Grid169/HeightBalanced.java
package Grid169;

public class HeightBalanced {

    static class Node{

    int data;
    Node left;
    Node right;
    Node(int data){
        this.data =data;
        this.left =null;
        this.right =null;
    }
    }

    static boolean isBalanced(Node node){

        if(node ==null)
            return true;

        int lh;
        int rh;

        lh = height(node.left);
        rh =height(node.right);

        if(Math.abs(lh -rh) <=1 && isBalanced(node.left) && isBalanced(node.right))
            return true;

        return false;

    }

    static int height(Node node){

        if(node == null)
            return 0;

        return 1 +  Math.max(height(node.left),height(node.right));

    }

    public static void main(String[] args){
        Node root =new Node(1);
        root.left =new Node(2);
        root.right =new Node(3);
        root.right.right =new Node(4);
        root.right.right.left =new Node(5);
        System.out.println(isBalanced(root));
    }
}
